File: src/main/java/com/cuupa/classificator/services/kb/semantic/token/Tokens.java
package com.cuupa.classificator.services.kb.semantic.token;

import org.apache.logging.log4j.util.Strings;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class Tokens {

	@Nullable
	public static Token get(@NotNull final TokenTextPointer tokenTextPointer) {
		String tokenName = findTokenName(tokenTextPointer);
		List<String> tokenValues = findTokenValue(tokenTextPointer);
		Token token = get(tokenName);
		token.setTokenValue(tokenValues);
		return token;
	}

	@Nullable
	private static Token get(String tokenName) {
		if ("oneOf".equals(tokenName)) {
			return new OneOf();
		} else if ("not".equals(tokenName)) {
			return new Not();
		} else if ("all".equals(tokenName)) {
			return new All();
		} else {
			return null;
		}
	}

	private static String findTokenName(@NotNull final TokenTextPointer pointer) {
		String tokenName = Strings.EMPTY;

		for (int i = pointer.getIndex() - 1; i > 0; i--) {
			if (pointer.get(i) != '{' && pointer.get(i) != ',' && pointer.get(i) > 64 && pointer.get(i) < 123) {
				tokenName = pointer.get(i) + tokenName;
			} else {
				return tokenName;
			}
		}
		return tokenName;
	}

	@NotNull
	private static List<String> findTokenValue(@NotNull final TokenTextPointer pointer) {
		List<String> value = new ArrayList<>();

		StringBuilder tokenValue = new StringBuilder();
		for (int i = pointer.getIndex() + 1; i < pointer.getCharSize(); i++) {
			if (pointer.get(i) == ',') {
				value.add(tokenValue.toString());
				tokenValue = new StringBuilder();
			} else if (pointer.get(i) != '"' && pointer.get(i) != ')') {
				tokenValue.append(pointer.get(i));
			} else if (pointer.get(i) == ')') {
				value.add(tokenValue.toString());
				return value;
			}
		}
		return value;
	}

	private static boolean isNextToken(@NotNull final TokenTextPointer pointer, final int i) {
		return pointer.get(i) == ')' && (pointer.get(i + 1) == ',' || pointer.get(i + 1) == '\r' || pointer.get(i + 1) == '\n');
	}
}

File: src/main/java/com/cuupa/classificator/services/kb/semantic/token/Tokens.kt
package com.cuupa.classificator.services.kb.semantic.token

import org.apache.logging.log4j.util.Strings

object Tokens {

    operator fun get(tokenTextPointer: TokenTextPointer): Token {
        val token = get(findTokenName(tokenTextPointer))
        token.tokenValue = findTokenValue(tokenTextPointer)
        return token
    }

    private operator fun get(tokenName: String): Token {
        return when (tokenName) {
            OneOf.name -> OneOf()
            Not.name -> Not()
            All.name -> All()
            WildcardBefore.name -> WildcardBefore()
            else -> Invalid()
        }
    }

    private fun findTokenName(pointer: TokenTextPointer): String {
        var tokenName = Strings.EMPTY
        for (i in pointer.index - 1 downTo 1) {
            tokenName = if (pointer[i] != '{' && pointer[i] != ',' && pointer[i].toInt() > 64 && pointer[i].toInt() < 123) {
                pointer[i].toString() + tokenName
            } else {
                return tokenName
            }
        }
        return tokenName
    }

    private fun findTokenValue(pointer: TokenTextPointer): MutableList<String> {
        val value: MutableList<String> = mutableListOf()
        var tokenValue = StringBuilder()
        for (i in pointer.index + 1 until pointer.charSize) {
            if (pointer[i] == ',') {
                value.add(tokenValue.toString())
                tokenValue = StringBuilder()
            } else if (pointer[i] != '"' && pointer[i] != ')') {
                tokenValue.append(pointer[i])
            } else if (pointer[i] == ')') {
                value.add(tokenValue.toString())
                return value
            }
        }
        return value
    }
}